# HW3 - Classification with Support Vector Machine
## Naive Handwriting Recognition System

#### Data Science in Korea University
#### COSE471, Spring 2015
#### Due : 5/24 (SUN)  11:59 PM 


## [ For Mac users, please make sure that your file runs on Windows! ]


## Preparing the Data

Put our traing examples and testing examples in your "IPython Notebook" folder or your original directory for your source
(train folder, test folder)




###We will go through the following steps:
* Read the files from folder
* Convert from the image(32X32) format to the list(1X1024) format
* Train your Training files with SVM library
* Predict Test files with your trained SVM model
* Measure your accuracy for your Test
* Compare with other Classificaiton Models ( 1 or 2 model)

To give you some starting points for your homework, we will also show the different sub-steps that can be taken to reach the presented solution.

## 0. Prepare your Data

Make sure your traning examples and testing examples in your "IPython Notebook" directory

In [1]:
import os
import numpy as np

## 1. Preprocess your Data

* Read the files from folder

* For Training Data / Convert from the image(32X32) format to the list(1X1024) format
* For Training Label(Answer for Data) / Name of the file

## 2. Train your Training files with SVM library

* Refer to this link
* http://scikit-learn.org/stable/modules/svm.html
* Try to set your parameter for SVM model

In [2]:
xandy = np.genfromtxt('/Users/hee/Practice_Python/DataScience/CreditScoring/training.csv',delimiter=',')

In [3]:
print xandy
print xandy.shape

xandy = xandy[1:,:]
print xandy.shape

y = xandy[:, 1]
x = xandy[:, 2:]
x.shape
y.shape

[[             nan              nan              nan ...,              nan
               nan              nan]
 [  4.20820000e+04   1.00000000e+00   1.17350929e+00 ...,   0.00000000e+00
    2.00000000e+00   2.00000000e+00]
 [  4.74630000e+04   0.00000000e+00   4.01181850e-02 ...,   0.00000000e+00
    0.00000000e+00   0.00000000e+00]
 ..., 
 [  5.75000000e+04   0.00000000e+00   8.17888577e-01 ...,   0.00000000e+00
    0.00000000e+00   0.00000000e+00]
 [  1.29582000e+05   1.00000000e+00   9.48217769e-01 ...,   2.00000000e+00
    0.00000000e+00   1.00000000e+00]
 [  9.01260000e+04   0.00000000e+00   1.12952920e-01 ...,   2.00000000e+00
    0.00000000e+00   0.00000000e+00]]
(90001, 12)
(90000, 12)


(90000,)

In [4]:
x = np.nan_to_num(x)
y = np.nan_to_num(y)

In [9]:
print x.shape
print y.shape

from sklearn.svm import SVC
xpart = x[:1000, :]
ypart = y[:1000]

print xpart.shape
print ypart.shape
#write your code here

# Train
clf = SVC(kernel = 'linear')
clf.fit(xpart, ypart)

(90000, 10)
(90000,)
(1000, 10)
(1000,)


SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0, degree=3, gamma=0.0,
  kernel='linear', max_iter=-1, probability=False, random_state=None,
  shrinking=True, tol=0.001, verbose=False)

## 3. Predict Test files with your trained SVM model

* Read the files from folder

* For Test Data / Convert from the image(32X32) format to the list(1X1024) format
* For Test Label(Answer for Data) / Name of the file

In [6]:
xTest = np.array(testList)
yTest = np.array(testAnswerList)

prediction = clf.predict(xTest)
print prediction

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 7 1 1 1 1 1 1 1 1 1 1 1 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 8 8 8 2 2 2 8 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 9 3 3 3 9 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3
 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 9 3 3 3 3 5 8 3 3 3 3 3 3 3 3 3 3 3 3
 9 3 3 3 3 3 3 3 3 3 3 3 3 9 3 3 3 3 3 3 3 3 3 3 3 3 3 3 4 4 4 4 4 4 4 4 4
 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 9 4 4 4 4 4 4 4 4 4 1 4 4 4 4 4 4 4 4 4
 4 4 4 4 4 4 4 4 4 0 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4
 4 4 4 4 4 4 4 4 4 4 4 4 8 4 4 4 4 4 8 4 4 4 4 4 4 4 4 4 4 4 4 5 5 5 5 5 5
 5 5 5 5 5 5 5 5 5 5 5 5 

## 4. Measure your accuracy for your Test Data

* Accuracy = (nubmer of correct answers) / (all number of documents)

In [7]:
#write your code here
accuracy = np.mean(prediction == yTest)
print accuracy

0.973572938689


## 5. Compare with other Classificaiton Models ( 1 or 2 model)

* Compare your results and Show your analysis for each results
* You can also analyse about different types of parameter for SVM
* You are welcome to use all kinds of libraries

##TABLE OF CONTENTS

###I. Logistic Regression
    1. Train
    2. Predict
    3. Accuracy
    4. Compare with SVM
    
###II. Modify parameters for SVM
    1. C
    2. kernel
    3. degree

##I. Logistic Regression

In [8]:
# 1. Train

from sklearn.linear_model import LogisticRegression

clfLR = LogisticRegression()
clfLR.fit(x, y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr',
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0)

In [9]:
# 2. Predict

predictionLR = clfLR.predict(xTest)
print predictionLR

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 9 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 7 1 1 1 1 1 1 1 1 1 1 1 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 8 8 8 3 2 2 8 8 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 9 3 3 3 9 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3
 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 9 3 3 3 3 5 8 3 3 3 3 3 3 3 3 3 3 3 3
 3 3 3 3 3 3 3 3 3 3 3 3 3 9 3 3 3 3 3 3 3 3 3 3 3 3 3 3 4 4 4 4 4 4 9 4 4
 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4
 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 1 4 4 4 4 4 4 4 4 4 4
 4 4 4 4 4 4 4 4 4 4 9 4 8 4 4 4 4 4 8 4 4 4 4 4 4 4 4 4 4 4 4 5 5 5 5 5 5
 5 5 5 5 5 5 5 5 5 5 5 5 

In [10]:
# 3. Accuracy

accuracy = np.mean(predictionLR == yTest)
print accuracy

0.966173361522


####4. Compare with SVM

    * Accuracy of SVM = 0.973572938689
    * Accuracy of Logistic Regression = 0.966173361522
    
    SVM is more accurate than Logistic Regression in this example.

##II. Modify Parameters for SVM

In [12]:
# 1. C
accuracyList = []
for i in np.linspace(1, 20, 20):
    clf = SVC(C = i)
    clf.fit(x, y)
    prediction = clf.predict(x)
    accuracy = np.mean(prediction == y)
    accuracyList.append(accuracy)
    print 'When C = %d, accuracy is %f.' % (i, accuracy)
    


When C = 1, accuracy is 0.979835.
When C = 2, accuracy is 0.988625.
When C = 3, accuracy is 0.992761.
When C = 4, accuracy is 0.994312.
When C = 5, accuracy is 0.996381.
When C = 6, accuracy is 0.997415.
When C = 7, accuracy is 0.998449.
When C = 8, accuracy is 0.998966.
When C = 9, accuracy is 0.998966.
When C = 10, accuracy is 0.998966.
When C = 11, accuracy is 0.999483.
When C = 12, accuracy is 1.000000.
When C = 13, accuracy is 1.000000.
When C = 14, accuracy is 1.000000.
When C = 15, accuracy is 1.000000.
When C = 16, accuracy is 1.000000.
When C = 17, accuracy is 1.000000.
When C = 18, accuracy is 1.000000.
When C = 19, accuracy is 1.000000.
When C = 20, accuracy is 1.000000.


In [ ]:
%matplotlib inline 
import matplotlib.pyplot as plt

plt.figure()
plt.plot(np.linspace(1, 20, 20), accuracyList, label = 'Accuracy')

plt.legend(loc='best')
plt.xticks(rotation=70) 
plt.xlabel("C")
plt.ylabel("Accuracy")
plt.title("C - Accuracy relationship")

With increasing C, accuracy increases as well, however to a certain level.

In [ ]:
# 2. kernel
accuracyListK = []
for i in ['linear', 'poly', 'rbf', 'sigmoid']:
    clf = SVC(kernel = i)
    clf.fit(x, y)
    prediction = clf.predict(xTest)
    accuracy = np.mean(prediction == yTest)
    accuracyListK.append(accuracy)
    print 'When kernel = %s, accuracy is %f.' % (i, accuracy)


From highest accuracy to lowest accuracy: linear > rbf > poly > sigmoid

With 'linear' kernel, accuracy was the highest.

In [ ]:
# 3. degree
accuracyListD = []
for i in range(1, 6):
    clf = SVC(kernel = 'poly', degree = i) # degree only applies when kernel is poly
    clf.fit(x, y)
    prediction = clf.predict(xTest)
    accuracy = np.mean(prediction == yTest)
    accuracyListD.append(accuracy)
    print 'When degree = %d, accuracy is %f.' % (i, accuracy)



In [ ]:
plt.figure()
plt.plot(range(1, 6), accuracyListD, label = 'Accuracy')

plt.legend(loc='best')
plt.xticks(rotation=70) 
plt.xlabel("Degree")
plt.ylabel("Accuracy")
plt.title("Degree - Accuracy relationship (kernel = 'poly')")

When using 'poly' kernel, degree 1 results in the highest accuracy.

## [ For Mac users, please make sure that your file runs on Windows! ] 
































